reactive/etcd.py

```diff
@@ -602,6 +608,50 @@ def render_default_user_ssl_exports():
     set_state('etcd.ssl.exported')
 
 
+def force_rejoin():
+    """Wipe local data and rejoin new cluster formed by leader unit
+
+    This action is required if leader unit performed snapshot restore. All
+    other members must remove their local data and previous cluster
+    identities and join newly formed, restored, cluster.
+    """
+    log('Wiping local storage and rejoining cluster')
+    conf = EtcdDatabag()
+    host.service_stop(conf.etcd_daemon)
+    clear_flag('etcd.registered')
+    etcd_data = os.path.join(conf.storage_path(), 'member')
+    if os.path.exists(etcd_data):
+        shutil.rmtree(etcd_data)
+    for _ in range(11):
+        # We need randomized back-off timer because only one unit can be
+        # joining at the same time
+        time.sleep(random.randint(1, 10))
+        register_node_with_leader(None)
+        if is_flag_set('etcd.registered'):
+            log('Successfully rejoined the cluster')
+            # Bump other members to run check_cluster_health and update status
+            hookenv.relation_set(hookenv.peer_relation_id(),
+                                 cluster_update=uuid4())
```

Comment:
To add on to this, the actual reason why you're seeing partial cluster health reports is because as soon as a given unit successfully re-registers itself, it will check the cluster status at that point in time, report it, and then exit the hook.  Given that all units will be running the rebuild / rejoin around the same time, with randomness introduced both by Juju, the cloud environment, and the random sleep, there's plenty of window for a given unit to see the status in a partial state at the point at which it registers itself. Furthermore, in addition to the hook thrashing that George mentioned, there's still a window of opportunity where the units may see partial status, even with the extra hook calls (although the chance of that happening would be lessened due to the extra delay caused by the additional hooks).

I was about to say that it might be worth having the units loop on the cluster health until everything comes back up, but that wouldn't work for the leader because the leader data won't be finalized until the action context ends, so it would be waiting for the other units to respond to a message that it hasn't told Juju it's good to send yet. So I guess the options are either accept the stale status, or do a proper coordinated restart using [the coordinator layer](https://git.launchpad.net/layer-coordinator).  That does get back into the realm of extra hooks, though.

A third option, if you really wanted to fix this properly, would be for the charm to register a service that runs in the background and checks the cluster health, and uses `juju-run` to invoke `update-status` if the health status changes.  This way, it would be responsive regardless of what the hook pattern looked like and would never rely on `update-status` and the 5-minute window.

Reply:
All of that said, in the end I agree with George, and it's probably best to just wait for `update-status` and not worry about it.  It's been suggested that the unit status really isn't the proper place for reporting info like this anyway and that `active` status really shouldn't have *any* sort of message.  Instead, a proper monitoring stack should be put in place for the cluster, and the unit status would just be used for monitoring the state of the charm itself.

Reply:
I would say that the "thrashing" like this is not the worst thing during the explicit operation like `restore` where high load is to be expected anyway and it's not triggered unexpectedly during normal operation. But since you both seem to agree on this, I'll remove cluster health bumping.